### PR TITLE
Support client country property with minor code cleanup

### PR DESCRIPTION
**Pull request by collina**
In addition, covers tickets: #23

### CODE REVIEW

**Comment by mtlynch** [comment]
Can you update documentation/selector-file-spec.md as well? We'll need to do it eventually and it's helpful to have it in this pull request so I understand the changes at a high level.

File: telescope/selector.py [hunk 1]
```diff
@@ -269,6 +286,15 @@ def validate_common(self, selector_dict):
         if not selector_dict.has_key('duration'):
             raise ValueError('UnsupportedDuration')
 
+        if selector_dict.has_key('client_countries'):
+            for client_country in selector_dict['client_countries']:
+                if len(client_country) != 2:
+                    raise ValueError('Requires ISO alpha-2 country code.')
+        if not selector_dict.has_key('client_countries') and \
```
Comment (mtlynch) [inline telescope/selector.py]: Style guide says to use parentheses for implied line continuation instead of backslash
https://google-styleguide.googlecode.com/svn/trunk/pyguide.html?showone=Line_length#Line_length

File: telescope/query.py [hunk 2]
```diff
@@ -169,15 +176,23 @@ def _create_query_string(self, mlab_project='ndt'):
...
+    
+    if self._conditional_dict.has_key('client_ips'):
+        client_ips_join = " OR\n\t\t".join(self._conditional_dict['client_ips'])
+        conditional_list_string += "\n\tAND ({client_ips})".format(
+            client_ips = client_ips_join)
+    
+    if self._conditional_dict.has_key('client_country'):
+        conditional_list_string += "\n\tAND ({conditional})".format(
+            conditional = self._conditional_dict['client_country'])
+    if self._conditional_dict.has_key('server_country'):
```
Comment (mtlynch) [inline telescope/query.py]: I'm confused about `server_country` because I don't see it in the selector. Is this only partially implemented at this point?

File: telescope/selector.py [hunk 3]
```diff
@@ -85,24 +85,36 @@ def __init__(self):
...
         selectors.append(selector)
+
+        selector.client_provider = client_provider
+        selector.client_country = client_country
+        selector.site = site
+        
+        if selector.client_provider != None:
+            selector.client_provider = selector.client_provider.lower()
+        if selector.client_country != None:
+            selector.client_country = selector.client_country.lower()
```
Comment (mtlynch) [inline telescope/selector.py]: We should add `client_country` to `Selector.__init__` so that it's obvious what attributes a `Selector` object should have.

File: telescope/selector.py [hunk 4]
```diff
@@ -85,24 +85,36 @@ def __init__(self):
...
-        selector.site = site
         selector.metric = metric
         selector.mlab_project = _metric_to_project[metric]
         selectors.append(selector)
+
+        selector.client_provider = client_provider
+        selector.client_country = client_country
+        selector.site = site
+        
+        if selector.client_provider != None:
```
Comment (mtlynch) [inline telescope/selector.py]: I think we should do the normalization in the parser rather than here.

File: telescope/selector_test.py [hunk 5]
```diff
@@ -40,8 +40,18 @@ def assertSelectorMatches(self, selector_expected, selector_actual):
     self.assertEqual(selector_expected.ip_translation_spec.strategy_name,
                      selector_actual.ip_translation_spec.strategy_name)
     self.assertDictEqual(selector_expected.ip_translation_spec.params, selector_actual.ip_translation_spec.params)
-    self.assertEqual(selector_expected.site, selector_actual.site)
-    self.assertEqual(selector_expected.client_provider, selector_actual.client_provider)
+
+    if hasattr(selector_expected, 'site') or \
+            selector_actual.site != None:
+        self.assertEqual(selector_expected.site, selector_actual.site)
```
Comment (mtlynch) [inline telescope/selector_test.py]: Do we need to do the `if`? Won't every `Selector` have a `site` attribute by virtue of the `__init__` method? Given that, it seems like we can do the `assertEqual` without any pre-check.

File: telescope/selector_test.py [hunk 6]
```diff
@@ -126,6 +137,7 @@ def testValidInput_v1dot1_Simple(self):
                                                  {'db_snapshots': ['2014-08-04']}))
     selector_expected.site = 'lga02'
     selector_expected.client_provider = 'comcast'
+    selector_expected.client_country = 'us'
```
Comment (mtlynch) [inline telescope/selector_test.py]: I think we need several more test cases here because the selector format is significantly more complex now that we have fields that can be either / or / both. What are the rules exactly with sites, providers, and countries? Is the user allowed to specify just one of sites, providers, and countries?

File: telescope/query.py [hunk 7]
```diff
@@ -257,9 +270,19 @@ def _add_server_ips_conditional(self, server_ips, is_web100):
...
     for server_ip in unique_server_ips:
       new_statement = '{local_ip_fieldname} = \'{server_ip}\''.format(
           local_ip_fieldname=local_ip_fieldname, server_ip=server_ip)
-      self._conditional_dict['server_ip'].append(new_statement)
-
+      self._conditional_dict['server_ips'].append(new_statement)
+
+  def _add_client_country_conditional(self, client_country):
+    self._conditional_dict['client_country'] = ("{fieldname} = " +
+        "UPPER('{client_country}')").format(fieldname =
```
Comment (mtlynch) [inline telescope/query.py]: Given that we explicitly call `lower()` on `client_country` in selector.py...

image

Seriously though, why don't we just uppercase the country code at parse time?

File: telescope/query.py [hunk 8]
```diff
@@ -169,15 +176,23 @@ def _create_query_string(self, mlab_project='ndt'):
...
+            server_ips = server_ips_join)
+    
+    if self._conditional_dict.has_key('client_ips'):
+        client_ips_join = " OR\n\t\t".join(self._conditional_dict['client_ips'])
+        conditional_list_string += "\n\tAND ({client_ips})".format(
+            client_ips = client_ips_join)
+    
+    if self._conditional_dict.has_key('client_country'):
+        conditional_list_string += "\n\tAND ({conditional})".format(
+            conditional = self._conditional_dict['client_country'])
```
Comment (mtlynch) [inline telescope/query.py]: We're repeating a lot of the same code here. We should try to refactor to a helper funciton.

File: main.py [hunk 9]
```diff
@@ -46,7 +46,8 @@
 class NoClientNetworkBlocksFound(Exception):
   def __init__(self, provider_name):
     Exception.__init__(self,
-                       'Could not find IP blocks associated with client provider {client_provider}.'.format(
+                       ('Could not find IP blocks associated with client ' +
+                       'provider {client_provider}.').format(
```
Comment (mtlynch) [inline main.py]: The hanging indents for a lot of these changes aren't consistent with the style guide. See https://google-styleguide.googlecode.com/svn/trunk/pyguide.html?showone=Indentation#Indentation

File: telescope/query.py (hunk 7 above)
Comment (collina) [inline telescope/query.py]: I had chiefly put that there for consistency in the writing of filenames, but having stepped back, those such be handled in the build_filename function if we want it.

File: telescope/query.py (hunk 2 above)
Comment (collina) [inline telescope/query.py]: Yes, it was in a previous version but I didn't fully implement it here because I am not sure its necessary at this time and don't want to overexpand Telescope yet. I'll remove this and if it is of value, we could revisit the issue later.

File: telescope/selector_test.py (hunk 5 above)
Comment (collina) [inline telescope/selector_test.py]: True, and that is solved by all possible properties being included in the ``__init__``, which wasn't the case for countries before and caused issues I was overreacting to.

File: telescope/query.py (hunk 7 above)
Comment (collina) [inline telescope/query.py]: This is the only case where it needs to be uppercase, whereas in other cases (e.g. filenames) it should be lower or it doesn't matter, so I am going to normalize them to lower in the Selector parser and then pass the value to the query builder with .upper() beforehand.

File: main.py (hunk 9 above)
Comment (collina) [inline main.py]: On the GVC I mentioned I would defer styleguide updates until the next pull, but I want to vet that my understanding on how this should properly be indented is correct, so I've modified it.

File: telescope/query.py (hunk 8 above)
Comment (collina) [inline telescope/query.py]: The `*_ips` and `client_country` are a bit different, so I don't think it's super repetitive, I'd invite ideas about how to do this without making is overcomplex for a couple of additional lines.

File: documentation/selector-file-spec.md [hunk 10]
```diff
@@ -8,26 +8,27 @@ The data subset selector file specifies subsets of M-Lab data, either for single
...
+    "sites": ["lga01", "lga02"],
+    "client_providers": ["Verizon", "comcast", "cablevision"],
+    "client_countries": ["us"]
 }
 ```
 
 # Field Specifications
 
-`file_format_version`: Specifies the version of this format file. This must be 1.
+`file_format_version`: Specifies the version of this format file. This must be 1.1
```
Comment (mtlynch) [inline documentation/selector-file-spec.md]: Missing ending period.

File: telescope/selector.py [hunk 11]
```diff
@@ -269,8 +286,13 @@ def validate_common(self, selector_dict):
         if not selector_dict.has_key('duration'):
             raise ValueError('UnsupportedDuration')
 
-        if not selector_dict.has_key('metrics') or \
-            type(selector_dict['metrics']) != list:
+        if selector_dict.has_key('client_countries'):
+            for client_country in selector_dict['client_countries']:
+                if len(client_country) != 2:
```
Comment (mtlynch) [inline telescope/selector.py]: Suggest 

```
if not re.match('^[a-zA-Z]{2}$', client_country):
```

File: documentation/selector-file-spec.md [hunk 12]
```diff
@@ -8,26 +8,27 @@ The data subset selector file specifies subsets of M-Lab data, either for single
...
 
 `duration`: Duration of time window (in days). Value must end with 'd'.
 
-`metric`: Text name of the metric for which to gather data. Valid values are:
+`metrics`: List of metrics for which to gather data. Valid values are:
 * `all` - Retrieves data for all metrics.
```
Comment (mtlynch) [inline documentation/selector-file-spec.md]: `all` is no longer supported, right?

File: documentation/selector-file-spec.md [hunk 13]
```diff
@@ -44,16 +45,26 @@ The data subset selector file specifies subsets of M-Lab data, either for single
...
-`start_time`: Start time of the window in which to collect test results (in ISO 8601 format). This value must end in `Z` (i.e. only UTC time zone is supported).
+`client_countries` _(optional)_: The ISO 3166-1 alpha-2 country code associated with the IP address of the measurement client. The geolocation of the client address is recorded within BigQuery at the time of measurement based on MaxMind current to that date.
+
+# Changlog 
+
+## As of version 1.1
+
+* Added `client_countries` property.
+* The properties `metric`, `client_provider` and `site` are now represented by the lists `metrics`, `client_providers` and `sites`. 
+* Made `client_providers`, `client_countries`, and `sites` optional.
```
Comment (mtlynch) [inline documentation/selector-file-spec.md]: (minor, nitpick) `client_countries` is new so its optional property didn't really change.

File: main.py [hunk 14]
```diff
@@ -176,7 +175,8 @@ def write_metric_calculations_to_file(data_filepath, metric_calculations, should
   return False
 
 
-def build_filename(resource_type, outpath, date, duration, site, client_provider, metric):
+def build_filename(resource_type, outpath, date, duration, site,
+                    client_provider, client_country, metric):
```
Comment (mtlynch) [inline main.py]: Can we move this to utils.py and add tests? I'd like to trim down main.py as much as possible and reduce the number of functions that are missing unit test coverage.

File: main.py [hunk 15]
```diff
@@ -205,19 +208,20 @@ def build_filename(resource_type, outpath, date, duration, site, client_provider
        (str): The generated full pathname of the output file.
   """
   extensions = { 'data': 'raw.csv', 'bigquery': 'bigquery.sql'}
-  filename_format = "{date}+{duration}_{site}_{client_provider}_{metric}-{extension}"
-
+  filename_format = "{date}+{duration}_{location_properties}_{metric}-{extension}"
+  
+  location_properties = "_".join(filter(None, [site, client_country,
```
Comment (mtlynch) [inline main.py]: (optional) What do you think of just `additional_properties`? `location_properties` seems wrong to me as a category for client ISP.

File: main.py [hunk 16]
```diff
@@ -479,25 +490,27 @@ def main(args):
...
+        telescope.utils.check_for_valid_cache(data_filepath) is True):
+      logger.info(('Raw data file found ({0}), assuming this is cached copy ' +
+                    'of same data and moving off. Use --ignorecache to ' +
+                    'suppress this behavior.').format(data_filepath))
       continue
-
-    logger.debug('Did not find existing data file: {data_filepath}'.format(**thread_metadata))
-    logger.debug(('Generating Query for subset of {site}, {client_provider}, {date}, ' +
-                  '{duration}.').format(**thread_metadata))
+
```
Comment (mtlynch) [inline main.py]: Trailing whitespace.

File: main.py [hunk 17]
```diff
@@ -64,6 +64,7 @@ def __init__(self):
     self.result = False
     self.metadata = None
     self.fatal_error = None
+    self.data_filepath = None
```
Comment (mtlynch) [inline main.py]: Can we document these attributes? Glancing at it, I'm realizing I can't figure out what any of them mean and can't infer the likely data types.

File: main.py [hunk 18]
```diff
@@ -46,8 +46,8 @@
 class NoClientNetworkBlocksFound(Exception):
   def __init__(self, provider_name):
     Exception.__init__(self,
-                       'Could not find IP blocks associated with client provider {client_provider}.'.format(
-                           client_provider = provider_name))
+        ('Could not find IP blocks associated with client provider '+
+        '{client_provider}.').format(client_provider = provider_name))
```
Comment (mtlynch) [inline main.py]: Don't need the `+` because Python will join the strings due to the parens. Need one more leading space on 50 to line up the quote characters:
https://google-styleguide.googlecode.com/svn/trunk/pyguide.html?showone=Line_length#Line_length

File: main.py [hunk 19]
```diff
@@ -298,27 +302,33 @@ def generate_query(selector, ip_translator, mlab_site_resolver):
...
-                                                             mlab_project = selector.mlab_project):
-      server_ips.append(retrieved_site_ip)
-      logger.debug("Found IP for {site} of {site_ip} on test {test}.".format(
-          site=selector.site, site_ip = retrieved_site_ip, test = selector.mlab_project))
-  except Exception as caught_error:
-    raise MLabServerResolutionFailed(caught_error)
-
+  if selector.client_provider != None:
+    client_ips = ip_translator.find_ip_blocks(selector.client_provider)
+    if len(client_ips) == 0:
```
Comment (mtlynch) [inline main.py]: Style guide likes implicit false:
https://google-styleguide.googlecode.com/svn/trunk/pyguide.html?showone=True/False_evaluations#True/False_evaluations

i.e. `if not client_ips:`

File: main.py [hunk 20]
```diff
@@ -298,27 +302,33 @@ def generate_query(selector, ip_translator, mlab_site_resolver):
...
+        raise NoClientNetworkBlocksFound(selector.client_provider)
+  else:
+    client_ips = None
```
Comment (mtlynch) [inline main.py]: For `client_ips` and `server_ips` we're currently making it either `None` or a non-empty list. Can we simplify the code by making these always be lists and they're either 0 or non-empty. I think we can avoid explicit checks for `None` before each operation and make it so that the loops just do nothing when the list is empty. i.e., it would be like:

```
client_ips = []
if selector.client_provider != None:
    client_ips = ip_translator.find_ip_blocks(selector.client_provider)
    if not client_ips:
        raise NoClientNetworkBlocksFound(selector.client_provider)
```

File: main.py [hunk 21]
```diff
@@ -298,27 +302,33 @@ def generate_query(selector, ip_translator, mlab_site_resolver):
...
+  
+  if selector.site != None:
+      server_ips = []
+      try:
+        for retrieved_site_ip in mlab_site_resolver.get_site_ips(selector.site,
+                                                                 mlab_project = selector.mlab_project):
+          server_ips.append(retrieved_site_ip)
+          logger.debug("Found IP for {site} of {site_ip} on test {test}.".format(
+              site=selector.site, site_ip = retrieved_site_ip, test = selector.mlab_project))
```
Comment (mtlynch) [inline main.py]: No whitespace around the `=` (they should all be like `site=selector.site`)
https://google-styleguide.googlecode.com/svn/trunk/pyguide.html?showone=Whitespace#Whitespace

File: main.py [hunk 22]
```diff
@@ -479,25 +490,27 @@ def main(args):
...
                     }
-    thread_metadata['data_filepath'] = build_filename('data',
-                                                      args.output,
-                                                      thread_metadata['date'],
-                                                      thread_metadata['duration'],
-                                                      thread_metadata['site'],
-                                                      thread_metadata['client_provider'],
-                                                      thread_metadata['metric'])
+    data_filepath = build_filename('data',
+                                      args.output,
```
Comment (mtlynch) [inline main.py]: Should be aligned with the opening paren:
https://google-styleguide.googlecode.com/svn/trunk/pyguide.html?showone=Indentation#Indentation

File: telescope/external.py [hunk 23]
```diff
@@ -200,38 +200,35 @@ def monitor_query_queue(self, job_id, job_metadata, query_object = None, callbac
...
         try:
           job_collection = query_object.authenticated_service.jobs()
           job_collection_state = job_collection.get(projectId = self.project_id, jobId = job_id).execute()
         except (SSLError, Exception, AttributeError, HttpError, httplib2.ServerNotFoundError) as caught_error:
           self.logger.warn(('Encountered error ({caught_error}) monitoring ' +
-                            'for {notification_identifier}, could be temporary, ' +
-                            'not bailing out.').format(caught_error = caught_error,
-                                                      notification_identifier = notification_identifier))
+                            'for ' + identifier_string  + ', could be temporary, ' +
+                            'not bailing out.').format(caught_error = caught_error))
```
Comment (mtlynch) [inline telescope/external.py]: Seems weird to do concatenation in the middle of a string we're already calling `format` on. Can we keep it as a format param?

File: main.py [hunk 24]
```diff
@@ -298,27 +302,33 @@ def generate_query(selector, ip_translator, mlab_site_resolver):
...
-    for retrieved_site_ip in mlab_site_resolver.get_site_ips(selector.site,
-                                                             mlab_project = selector.mlab_project):
-      server_ips.append(retrieved_site_ip)
-      logger.debug("Found IP for {site} of {site_ip} on test {test}.".format(
-          site=selector.site, site_ip = retrieved_site_ip, test = selector.mlab_project))
-  except Exception as caught_error:
-    raise MLabServerResolutionFailed(caught_error)
-
+  if selector.client_provider != None:
+    client_ips = ip_translator.find_ip_blocks(selector.client_provider)
```
Comment (mtlynch) [inline main.py]: I think the `_blocks` suffix is useful because it indicates to the reader that these are not going to be like `['1.2.3.4', '2.3.4.5']` and instead they're `[(12345, 12348), ...]`. It's confusing if we name it `client_ips` and `server_ips` when they store very different data.

File: main.py [hunk 25]
```diff
@@ -479,25 +490,27 @@ def main(args):
                       'duration': duration_to_string(selector.duration),
                       'site': selector.site,
                       'client_provider': selector.client_provider,
-                      'metric': selector.metric,
-                      'mlab_project': selector.mlab_project,
+                      'client_country': selector.client_country,
+                      'metric': selector.metric
```
Comment (mtlynch) [inline main.py]: (optional): This is out of scope of your change, but do we actually need this `thread_metadata` dictionary? It's just a `Selector` with certain fields converted to strings. It would be nice to pass a `Selector` around instead and trim down the parameter list of a lot of these 8+ parameter functions.

File: telescope/query.py [hunk 26]
```diff
@@ -257,9 +267,14 @@ def _add_server_ips_conditional(self, server_ips, is_web100):
...
     for server_ip in unique_server_ips:
       new_statement = '{local_ip_fieldname} = \'{server_ip}\''.format(
           local_ip_fieldname=local_ip_fieldname, server_ip=server_ip)
-      self._conditional_dict['server_ip'].append(new_statement)
+      self._conditional_dict['server_ips'].append(new_statement)
 
+  def _add_client_country_conditional(self, client_country):
+    self._conditional_dict['client_country'] = ("{fieldname} = " +
+        "'{client_country}'").format(fieldname =
+        "connection_spec.client_geolocation.country_code",
```
Comment (mtlynch) [inline telescope/query.py]: I think this should just be part of the format string since it's a static string and splitting it out this way adds complexity.

File: telescope/query_test.py [hunk 27]
```diff
@@ -148,15 +154,58 @@ def test_ndt_download_throughput_query_full_month(self):
...
+  AND web100_log_entry.snap.CongSignals IS NOT NULL
+  AND web100_log_entry.connection_spec.remote_ip IS NOT NULL
+  AND web100_log_entry.connection_spec.local_ip IS NOT NULL
+  AND project = 0
+  AND web100_log_entry.is_last_entry = True
+  AND connection_spec.data_direction == 1
+  AND ((web100_log_entry.log_time >= 1388534400) AND (web100_log_entry.log_time < 1391212800))
+  AND (web100_log_entry.connection_spec.local_ip = '1.1.1.1' OR
+       web100_log_entry.connection_spec.local_ip = '2.2.2.2')
+  AND (connection_spec.client_geolocation.country_code = UPPER('us'))
```
Comment (mtlynch) [inline telescope/query_test.py]: This test is failing now right? In case you don't already have it set up, I highly recommend setting up the pre-commit hook so that you don't have to remember to run unit tests manually before committing:
https://github.com/m-lab-tools/telescope/wiki/Running-unit-tests

File: telescope/query_test.py [hunk 28]
```diff
@@ -148,15 +154,58 @@ def test_ndt_download_throughput_query_full_month(self):
...
+"""
+    self.assertQueriesEqual(query_expected, query_actual)
+
+
```
Comment (mtlynch) [inline telescope/query_test.py]: Now that we've made several fields optional, I think we should have coverage for the following cases
 1. none of the optional fields are specified
 1. all of the optional fields are specified (I think that's `test_ndt_download_throughput_query_country_property` right now)
 1. for each optional field, only that field is specified and the others are not

I'm fine with stripping out some of the existing tests because I don't think there's a huge value in having two optional fields set, one unset.

File: main.py [hunk 29]
```diff
@@ -527,17 +541,18 @@ def main(args):
           run the query thus far (failed queries are pushed back to the end
           of the loop).
       """
-      selector_queue.put( (bq_query_string, bq_table_span, thread_metadata, False) )
+      selector_queue.put( (bq_query_string, bq_table_span, thread_metadata, \
```
Comment (mtlynch) [inline main.py]: Don't need the whitespace between the parens or the trailing slash.

File: telescope/query.py (hunk 8 above)
Comment (mtlynch) [inline telescope/query.py]: I think we can simplify if we collect the conditional clauses in a list and then join them together with "AND" strings at the end. I just did something really similar here:
https://github.com/mtlynch/bigquerygenerator/blob/master/generator.html#L278

File: telescope/selector.py [hunk 30]
```diff
@@ -156,13 +163,21 @@ def _parse_input_for_selectors(self, selector_json):
...
         selector_json['start_times'])
-    multi_selector.client_providers = selector_json['client_providers']
-    multi_selector.sites = selector_json['sites']
     multi_selector.metrics = selector_json['metrics']
     multi_selector.duration = self._parse_duration(selector_json['duration'])
     multi_selector.ip_translation_spec = self._parse_ip_translation(
         selector_json['ip_translation'])
-
+    
+    if selector_json.has_key('client_providers'):
```
Comment (mtlynch) [inline telescope/selector.py]: Style guide likes `if 'client_providers' in selector_json:`
https://google-styleguide.googlecode.com/svn/trunk/pyguide.html?showone=Default_Iterators_and_Operators#Default_Iterators_and_Operators

File: documentation/selector-file-spec.md [hunk 31]
```diff
@@ -44,16 +45,26 @@ The data subset selector file specifies subsets of M-Lab data, either for single
 
 `db_snapshots`: Specifies the snapshot dates (in YYYY-MM-DD format) of the MaxMind databases that are required to resolve IP addresses to providers. This field is optional. If not specified or specified as an empty list, consuming programs should use database snapshots closest in time to the snapshots specified and should suppress warnings to the user about missing snapshots.
 
-`site`: Name of M-Lab site where data was collected.
+`start_times`: List of start times of the window in which to collect test results (in ISO 8601 format). This value must end in `Z` (i.e. only UTC time zone is supported).
```
Comment (mtlynch) [inline documentation/selector-file-spec.md]: Suggest changing `This value` to `Start time values` for clarity.

File: documentation/selector-file-spec.md [hunk 32]
```diff
@@ -44,16 +45,26 @@ The data subset selector file specifies subsets of M-Lab data, either for single
...
+
+`client_providers` _(optional)_: List of client providers based on their name. This is the substring that should appear in all AS names when a mapping is performed from this parameter to corresponding ASes. For example, "Verizon" should match AS names "Verizon Online LLC", "MCI Communications Services, Inc. d/b/a Verizon Business", "Verizon Data Services LLC", etc. Supports a limited number of provider metanames that translate to all known AS name queries for that provider:
 
 * `twc`: Time Warner Cable
 * `centurylink`: CenturyLink
 * `level3`: Level 3 Communications
 * `cablevision`: Cablevision Communications
 
-`start_time`: Start time of the window in which to collect test results (in ISO 8601 format). This value must end in `Z` (i.e. only UTC time zone is supported).
+`client_countries` _(optional)_: The ISO 3166-1 alpha-2 country code associated with the IP address of the measurement client. The geolocation of the client address is recorded within BigQuery at the time of measurement based on MaxMind current to that date.
```
Comment (mtlynch) [inline documentation/selector-file-spec.md]: Should make it clear that this is a list.

File: telescope/selector_test.py [hunk 33]
```diff
@@ -126,8 +127,133 @@ def testValidInput_v1dot1_Simple(self):
...
+    selector_expected.start_time = utils.make_datetime_utc_aware(datetime.datetime(2014, 2, 1))
+    selector_expected.duration = 30 * 24 * 60 * 60
+    selector_expected.metric = 'average_rtt'
+    selector_expected.ip_translation_spec = (
+         iptranslation.IPTranslationStrategySpec('maxmind',
+                                                 {'db_snapshots': ['2014-08-04']}))
+    selector_expected.client_country = 'us'
+    self.assertParsedSingleSelectorMatches(selector_expected, selector_file_contents)
+
+
```
Comment (mtlynch) [inline telescope/selector_test.py]: Can we add a test case where one of the optional parameters is an empty list? e.g. `metrics: []`

Such a file seems to adhere to our spec, but I think this breaks our code now. I think we should treat an empty list the same as if they omitted the field.

File: telescope/selector_test.py [hunk 34]
```diff
@@ -126,8 +127,133 @@ def testValidInput_v1dot1_Simple(self):
...
+    selector_expected.start_time = utils.make_datetime_utc_aware(datetime.datetime(2014, 2, 1))
+    selector_expected.duration = 30 * 24 * 60 * 60
+    selector_expected.metric = 'average_rtt'
+    selector_expected.ip_translation_spec = (
+         iptranslation.IPTranslationStrategySpec('maxmind',
+                                                 {'db_snapshots': ['2014-08-04']}))
+    selector_expected.site = 'lga02'
+    self.assertParsedSingleSelectorMatches(selector_expected, selector_file_contents)
+
+    selector_file_contents = """{
```
Comment (mtlynch) [inline telescope/selector_test.py]: Can you put these in separate functions? It helps reduce the possibility of tests polluting each other's results.

File: telescope/selector_test.py [hunk 35]
```diff
@@ -198,6 +324,7 @@ def testEncodeMultiSelectorOneElement(self):
     s.duration = 45
     s.site_names = ['mia01']
     s.client_providers = ['twc']
+    s.client_countries = ['us', 'ca', 'uk']
```
Comment (mtlynch) [inline telescope/selector_test.py]: This is my fault for terrible naming, but the idea of this test is that for every field in the multiselector whose type is a list, the list only contains one element.

Can you change this to `s.client_countries = ['us']` and move this line to `testEncodeMultiSelectorMultiElement` and if you can think of a better test name that makes the intent clearer, please change it.
